### PR TITLE
Fix #10508 (modernize virtualenv doc)

### LINKED ISSUE #10508: Modernize documentation on how to use virtualenv
### Description
The page on [virtualenv](https://docs.astropy.org/en/latest/development/workflow/virtualenv_detail.html) in the development documentation (which I believe I might have originally written most of), while still a good thing to have, is pretty outdated and should be cleaned up at some point.  A few points to highlight

* [x] In the usage instructions it still mentions the `--distribute` flag which has long been deprecated and is no longer relevant.
* [x] At the time of its original writing I don't think conda existed yet, or was not widely used.  For most beginners I would recommend conda (which we do elsewhere in the documentation) but it would be good here to clarify the relationship between conda environments and virtualenv environments (and why the former precludes use of the latter).
* [x] It would also be good to mention `venv` more prominently since it is more widely used these days, but also clarify its relationship with `virtualenv` (and why the latter is more flexible), but the former is also still useful in a pinch.
* [x] Maybe some mention of [Pipenv](https://pipenv.pypa.io/en/latest/)?
* [x] Clean up most of the documentation on using virtualenv with IPython; although there already is a notice to this effect, most of this documentation is long obsolete and would better be removed to reduce confusion.  Still, some note on how to use IPython/jupyter notebooks in a virtualenv would be good to have. 

**Update:** It turns out there is also already a newer page on working with different types of virtual environments: https://docs.astropy.org/en/latest/development/workflow/virtual_pythons.html#virtual-envs  I would probably just integrate I listed above that's missing from that page, and then either remove the old page or change it to a redirect to the new one.

### PR DESCRIPTION
**Pull request by embray**
* Remove old, outdated virtualenv documentation page and shift some details to the newer page.

* Added details on pipenv, as well as a link to the Hitchhiker's Guide to Python chapter on the subject, which was helpful for writing the pipenv docs.  Other minor cleanup to that page.

* Added Pipfile[.lock] to .gitignore for now, since we don't currently include a Pipfile in the source, though it is created by default when installing a package with pipenv.

Fix #10508

### CODE REVIEW

**Comment by embray** [comment]
Fixed undefined label.

**Comment by pllim** [comment]
I am changing the milestone to 4.0.x given this doc is very old, but if backport is a pain, release manager can change the milestone (again).

File: docs/development/workflow/virtual_pythons.rst
Comment (pllim) [inline docs/development/workflow/virtual_pythons.rst:218]: We made the distinction of using capitalized Astropy to indicate the whole project and `astropy` to indicate just this repo.

```suggestion
  directory (i.e., the copy of the ``astropy`` repository on your computer) and
```

File: docs/development/workflow/virtual_pythons.rst
Comment (pllim) [inline docs/development/workflow/virtual_pythons.rst:257]: This text here needs to mention `pipenv` too.

File: docs/install.rst
Comment (pllim) [inline docs/install.rst:155]: If this links to RST inside `development`, you might need to hardcode the URL due to #11342 . To be sure, you can tweak `docs/conf.py` to mimic a stable build locally and make sure that has no warnings. See #11344 for example.

**Review by pllim — COMMENTED** [review]
Thanks, nice clean-up! Some comments below.

I'll also ping someone who I think does not use `conda` to review, just to have extra pairs of eyes.

File: docs/development/workflow/virtual_pythons.rst
Comment (embray) [reply docs/development/workflow/virtual_pythons.rst:20]: Yes, I think that's a bit clearer.

File: docs/install.rst
Comment (embray) [reply docs/install.rst:155]: Here I was planning to rebase this on top of my fix for #11366.  For which I see I haven't made a PR yet; I thought I already did.

File: docs/development/workflow/virtual_pythons.rst
Comment (embray) [reply docs/development/workflow/virtual_pythons.rst]: I don't know, is that a policy?  It makes my eyes hurt a little when list items don't have blank lines between them.  I don't recall reST requiring that.

Comment (embray) [reply docs/development/workflow/virtual_pythons.rst]: Here's a screenshot of the relevant section as rendered:

Screenshot from 2021-03-15 02-54-22

Comment (pllim) [reply docs/development/workflow/virtual_pythons.rst]: As long as it doesn't break rendering, I guess it is fine. Sphinx has weird rules about line breaks in some places, so I just want to make sure. Thanks!

Comment (pllim) [reply docs/development/workflow/virtual_pythons.rst]: As long as it doesn't break rendering, I guess it is fine. Sphinx has weird rules about line breaks in some places, so I just want to make sure. Thanks!

File: docs/install.rst
Comment (pllim) [reply docs/install.rst:155]: Should I withhold approval until the rebase is done?

Comment (embray) [reply docs/install.rst]: Yes please, that would make sense.  (It is done now, see #11622 )

**Comment by embray** [comment]
This is rebased now on top of #11622, so it depends on that first.  But otherwise this is ready to go.

**Comment by pllim** [comment]
There is a conflict now, so let's wait till after #11622 is merged before we rebase and review.

**Comment by pllim** [comment]
#11622 is merged but there is still a conflict. I could resolve it for you but I am afraid it would mess up the commit authorship. Given this is milestoned to a bugfix release, it doesn't have to go in by 4.3 feature freeze. It just won't end up in 4.3 but will be in a later bugfix release, that's all. I'll leave it to follow-up when you can, unless you state otherwise. Thanks!

**Comment by embray** [comment]
Rebased.  No problem about missing the freeze.  Sorry I haven't been available lately.

File: docs/common_links.txt
Comment (pllim) [inline docs/common_links.txt:124]: Why is this line in a diff? I can't see any difference here!

File: docs/development/workflow/virtual_pythons.rst
Comment (pllim) [inline docs/development/workflow/virtual_pythons.rst:218]: While I realize `cd` is used in the example, it is not the only way to change a directory, and it doesn't work in native Windows shell, so I propose using normal English here.
```suggestion
* `pipenv`_: Activate the environment by changing into the project
```

File: docs/development/workflow/virtualenv_detail.rst
Comment (pllim) [inline docs/development/workflow/virtualenv_detail.rst:3]: Removing this ref might break some docs downstream. Example report along similar vein: #11754

**Review by pllim — COMMENTED** [review]
Some final nitpicks.

File: docs/common_links.txt
Comment (embray) [reply docs/common_links.txt:124]: Oh, I think my editor added a newline at the end of the line where there wasn't one before.  GitHub should show that!

File: docs/development/workflow/virtual_pythons.rst
Comment (embray) [reply docs/development/workflow/virtual_pythons.rst:218]: I'm not sure what you mean "it doesn't work in native Windows shell".  `cd` works fine in Windows.  I think it's probably the first DOS command I ever learned :)

Comment (embray) [reply docs/development/workflow/virtual_pythons.rst:218]: Fair point though that it's not technically the *only* way to change directories.  I often alias `cd` to `pushd` :)

Comment (pllim) [reply docs/development/workflow/virtual_pythons.rst:218]: Ops! I must have mixed up `cd` with `ls`. It does work on Powershell now that I check. My bad!

File: docs/development/workflow/virtualenv_detail.rst
Comment (embray) [reply docs/development/workflow/virtualenv_detail.rst:3]: Good point.  A quick search turned up at least one project that references that page: https://pypi.org/project/pywicta/0.1.dev19/

Good opportunity to try out sphinx-redirects.

File: docs/development/workflow/virtual_pythons.rst
Comment (embray) [reply docs/development/workflow/virtual_pythons.rst:218]: It's ok, I think your suggestion was good anyways.

File: docs/development/workflow/virtualenv_detail.rst
Comment (embray) [reply docs/development/workflow/virtualenv_detail.rst:3]: I added a redirect using sphinx-reredirects (I kept missing the "rere" in the name before).

@pllim Since this adds a new doc build dependency do you think I should make a separate PR for that first, or just go ahead and do it here since this is where the need arose?

File: setup.cfg
Comment (pllim) [inline setup.cfg:92]: I guess this package is from https://gitlab.com/documatt/sphinx-reredirects ? Are we worried about single-point-of-failure here, in that the author will abandon this package and then we end up with a bunch of broken links?

I think it is okay to introduce a new dependency here instead of a new PR. But I want to make sure we are not using someone's random pet project that will get abandoned.

Comment (pllim) [reply setup.cfg:92]: Perhaps a less controversial approach to let this PR get merged sooner than later is, as you said, introduce this dependency in a follow-up PR. In the meantime, instead of deleting the ref to `docs/development/workflow/virtualenv_detail.rst`, move that ref somewhere in the new page, or keep the old page with the ref, but scrub out the content and point to the new page without this reredirect thingy.

Comment (embray) [reply setup.cfg:92]: I think it's pretty low-risk.  The developer seems open to accepting patches, and the extension itself is about 100 lines so if it falls into disrepair, could easily be merged into sphinx-astropy or the like.

Otherwise I don't think this is worth expending any more effort on.

Comment (pllim) [reply setup.cfg:92]: Let's see what others say.

**Comment by embray** [comment]
I haven't seen anyone express a strong opinion on this yet, so for the sake of getting this done I'll revert the use of sphinx-reredirects for now and create a redirect in RTD before we merge this.

Then I'll open a separate issue for adding sphinx-reredirects

**Comment by embray** [comment]
I opened #11837 for the redirects stuff.  @pllim if it's ok with you I'll go ahead and merge this and set up a manual redirect for now?

**Review by mhvk — COMMENTED** [review]
Very small comment

**Review by pllim — COMMENTED** [review]
If you can set up the manual redirect for `using-virtualenv`, that is good enough for me. But I'll withhold approval until @mhvk is happy with the text. Thank you all for your patience!

**Review by mhvk — APPROVED** [review]
I think it is all good modulo my small grammatical comment. I'll approve pending that.

File: docs/development/workflow/virtual_pythons.rst
Comment (embray) [reply docs/development/workflow/virtual_pythons.rst]: After having another look at that paragraph I decided that the "as many packages as you want" phrase was maybe a bit confusing and unnecessary, so I just cut it out and fused that sentence with the next one.  Instead I added some follow-up detail on why you might want to use this.

Comment (embray) [reply docs/development/workflow/virtual_pythons.rst]: That's in 865e9c594

**Comment by mhvk** [comment]
Changes are all great, thanks! Is the manual redirect part of this, or separate? If separate, feel free to merge.

**Comment by embray** [comment]
Merging and setting up the manual redirect for now.

**Comment by embray** [comment]
I added the redirect but it doesn't seem to have taken effect yet.  In the past I seem to recall it being pretty quick.  I'll try again in a few minutes...

**Comment by embray** [comment]
OK, the redirect works now.

**Comment by pllim** [comment]
Thanks!

**Pull request merged**
